homeassistant/components/screenlogic/binary_sensor.py

```diff
@@ -1,161 +1,190 @@
+            },
+            GROUP.ALERT: {
+                VALUE.ORP_LIMIT: SupportedBinarySensorValueParameters(),
+                VALUE.PH_LIMIT: SupportedBinarySensorValueParameters(),
+                VALUE.PH_LOCKOUT: SupportedBinarySensorValueParameters(),
+            },
+            GROUP.WATER_BALANCE: {
+                VALUE.CORROSIVE: SupportedBinarySensorValueParameters(),
+                VALUE.SCALING: SupportedBinarySensorValueParameters(),
+            },
+        },
+        DEVICE.SCG: {
+            GROUP.SENSOR: {
+                VALUE.STATE: SupportedBinarySensorValueParameters(),
+            },
+        },
+    }
+)
 
 SL_DEVICE_TYPE_TO_HA_DEVICE_CLASS = {DEVICE_TYPE.ALARM: BinarySensorDeviceClass.PROBLEM}
 
-SUPPORTED_CONFIG_BINARY_SENSORS = (
-    "freeze_mode",
-    "pool_delay",
-    "spa_delay",
-    "cleaner_delay",
-)
-
 
 async def async_setup_entry(
     hass: HomeAssistant,
     config_entry: ConfigEntry,
     async_add_entities: AddEntitiesCallback,
 ) -> None:
     """Set up entry."""
-    entities: list[ScreenLogicBinarySensorEntity] = []
-    coordinator: ScreenlogicDataUpdateCoordinator = hass.data[DOMAIN][
+    entities: list[ScreenLogicBinarySensor] = []
+    coordinator: ScreenlogicDataUpdateCoordinator = hass.data[SL_DOMAIN][
         config_entry.entry_id
     ]
-    gateway_data = coordinator.gateway_data
-    config = gateway_data[SL_DATA.KEY_CONFIG]
-
-    # Generic binary sensor
-    entities.append(
-        ScreenLogicStatusBinarySensor(coordinator, "chem_alarm", CODE.STATUS_CHANGED)
-    )
-
-    entities.extend(
-        [
-            ScreenlogicConfigBinarySensor(coordinator, cfg_sensor, CODE.STATUS_CHANGED)
-            for cfg_sensor in config
-            if cfg_sensor in SUPPORTED_CONFIG_BINARY_SENSORS
-        ]
-    )
-
-    if config["equipment_flags"] & EQUIPMENT.FLAG_INTELLICHEM:
-        chemistry = gateway_data[SL_DATA.KEY_CHEMISTRY]
-        # IntelliChem alarm sensors
-        entities.extend(
-            [
-                ScreenlogicChemistryAlarmBinarySensor(
-                    coordinator, chem_alarm, CODE.CHEMISTRY_CHANGED
+    gateway = coordinator.gateway
+    data_path: ScreenLogicDataPath
+    value_params: SupportedBinarySensorValueParameters
+    for data_path, value_params in iterate_expand_group_wildcard(
+        gateway, SUPPORTED_DATA
+    ):
+        entity_key = generate_unique_id(*data_path)
+
+        device = data_path[0]
+
+        if not (DEVICE_INCLUSION_RULES.get(device) or value_params.included).test(
+            gateway, data_path
+        ):
+            cleanup_excluded_entity(coordinator, DOMAIN, entity_key)
+            continue
+
+        try:
+            value_data = gateway.get_data(*data_path, strict=True)
+        except KeyError:
+            _LOGGER.debug("Failed to find %s", data_path)
+            continue
+
+        entity_description_kwargs = {
+            **build_base_entity_description(
```

Comment:
The controllers allows you to build nearly any combination of circuits so we can't know in advance what the system will have.

Reply:
The goal with this was to be DRYer.  

I could conceivably fully define an entity description for every possible `binary_sensor` and `sensor` entity, but `SUPPORTED_DATA` would grow by ~4x in the number of lines of code in each `binary_sensor.py` and `sensor.py`; and it would not remove the need for processing of controller configuration to fill in missing required information in the entity descriptions.

If this is preferred though, I am happy to make that change. Let me know how you would like me to proceed.

Reply:
There should only be so many measurements. I'd look into breaking out the variable parts from the entity description and let the entity description capture the type of measurement with device class and unit of measurement.

Reply:
I initially looked at partially defining entity descriptions with common data, but it appeared I would have to initialize the defined `EntityDescription`s with `key=""`, which seemed undesirable given it is a required parameter.  Maybe that's not forbidden, though?

To your suggestion, there are 15 unique unit of measurements out of 25 `sensor` entities, with all UoMs provided by the api (`value_data`) hence the dynamic nature of the current setup. I could group device class and state class, and define an extra temperature device class entity description to account for a different entity category, which would reduce the count down to 7 partial entity descriptions, but they'll only have 3 points of data. 

Regardless, any arrangement short of defining every sensor's entity description will still need to explicitly define the data paths for all possible entities along with special parameters for any value adjustments and rules about when to include a sensor based on equipment configuration. That means `SUPPORTED_DATA` and most of the dynamic processing would need to stay the same, rather than be mostly replaced with the explicit entity descriptions. I fear that only partially defining entity descriptions would make the setup more verbose _and_ more complicated, rather than trading one for the other.

IMO, if code clarity is the priority, I'd vote for explicitly defining all entity descriptions outright.  That said, I am at your discretion.

Reply:
As long as there are not hundreds of descriptions required, there's no problem defining them explicitly. If there are more, we need to use a different approach where we pick a description from a smaller common set.